File: src/main/kotlin/dev/loveeev/astraTowny/data/unload/UnLoad.java
package dev.loveeev.astraTowny.data.unload;

import dev.loveeev.astraTowny.data.Cache;
import dev.loveeev.astratowny.objects.townblocks.WorldCoord;
import dev.loveeev.astratowny.objects.townblocks.TownBlocks;
import dev.loveeev.astratowny.database.SQL;
import dev.loveeev.astratowny.database.SchemeSQL;
import dev.loveeev.astratowny.manager.TownManager;
import dev.loveeev.astratowny.objects.Nation;
import dev.loveeev.astratowny.objects.Resident;
import dev.loveeev.astratowny.objects.Rank;
import dev.loveeev.astratowny.objects.Town;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.HashMap;
import java.util.Map;
import java.util.stream.Collectors;

public class UnLoad {
    public UnLoad() {
        if (Boolean.TRUE.equals(SQL.isCon())) {
            sendDataToDatabase();
        }
    }


    private void sendDataToDatabase() {
        TownManager.getInstance().getNations().values().parallelStream().forEach(this::sendNationsToDatabase);
        TownManager.getInstance().getRanks().values().parallelStream().forEach(this::sendRanksToDatabase);
        processBatch(TownManager.getInstance().getTownBlocks());
        TownManager.getInstance().getTowns().values().parallelStream().forEach(this::sendTownDataToDatabase);
        TownManager.getInstance().getResidents().values().parallelStream().forEach(this::sendResidentDataToDatabase);
    }

    private void sendNationsToDatabase(Nation nation) {
        try (Connection connection = SQL.getCon();
             PreparedStatement statement = connection.prepareStatement("REPLACE INTO " + SchemeSQL.TABLE_PREFIX + "NATIONS (uuid, name, capital, king, balance, mapcolor) VALUES (?, ?, ?, ?, ?, ?)")) {
            statement.setString(1, nation.getUuid().toString());
            statement.setString(2, nation.getName());
            statement.setString(3, nation.getCapital() != null ? nation.getCapital().getName() : null);
            statement.setString(4, nation.getKing() != null ? nation.getKing().getPlayerName() : null);
            statement.setDouble(5, nation.getBalance());
            statement.setString(6, nation.getMapColor() != null ? nation.getMapColor() : "000000");
            statement.executeUpdate();
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    private void sendRanksToDatabase(Rank rank) {
        try (Connection connection = SQL.getCon();
             PreparedStatement statement = connection.prepareStatement(
                     "REPLACE INTO " + SchemeSQL.TABLE_PREFIX + "RANKS (uuid, name, town, nation, perms) VALUES (?, ?, ?, ?, ?)")) {

            statement.setString(1, rank.getUuid().toString());
            statement.setString(2, rank.getName());

            if (rank.getTown() != null) {
                statement.setString(3, rank.getTown().getName());
                statement.setString(4, null); // Если есть город, нация будет NULL
            } else if (rank.getNation() != null) {
                statement.setString(3, null); // Если есть нация, город будет NULL
                statement.setString(4, rank.getNation().getName());
            } else {
                throw new IllegalArgumentException("Both town and nation are null for rank: " + rank.getName());
            }

            statement.setString(5, rank.getPermissionList() != null ?
                    "[" + rank.getPermissionList().stream().toList().toString().replace("[", "").replace("]", "") + "]" : null);

            statement.executeUpdate();

        } catch (SQLException e) {
            e.printStackTrace();
        }
    }



    private void processBatch(Map<WorldCoord, TownBlocks> batch) {
        try {
            batch.entrySet().parallelStream().forEach(entry -> {
                try (Connection connection = SQL.getCon();
                     PreparedStatement statement = connection.prepareStatement("REPLACE INTO " + SchemeSQL.TABLE_PREFIX + "TOWNBLOCKS (world, X, Z, town) VALUES (?, ?, ?, ?)")) {

                    TownBlocks value = entry.getValue();
                    statement.setString(1, value.getWorld() != null ? value.getWorld().getName() : null);
                    statement.setInt(2, value.getX());
                    statement.setInt(3, value.getZ());
                    statement.setString(4, value.getTown() != null ? value.getTown().getName() : null);
                    statement.executeUpdate();

                } catch (SQLException e) {
                    e.printStackTrace();
                }
            });
        } catch (Exception e) {
            // Handle exception here
            e.printStackTrace();
        }
    }


    private void sendTownDataToDatabase(Town town) {
        try (Connection connection = SQL.getCon();
             PreparedStatement statement = connection.prepareStatement("REPLACE INTO " + SchemeSQL.TABLE_PREFIX + "TOWNS (uuid, name, mayor, nation, homeblock, spawn, mapcolor) VALUES (?, ?, ?, ?, ?, ?, ?)")) {
            statement.setString(1, town.getUuid().toString());
            statement.setString(2, town.getName());
            statement.setString(3, town.getMayor() != null ? town.getMayor().getPlayerName() : null);
            statement.setString(4, town.getNation() != null ? town.getNation().getName() : null);
            statement.setString(5, town.getHomeblock() != null ? town.getHomeblock().getWorld().getName() + "#" + town.getHomeblock().getX() + "#" + town.getHomeblock().getZ() : null);
            statement.setString(6, town.getSpawnLocation() != null ? town.getSpawnLocation().getWorld().getName() + "#" + town.getSpawnLocation().getX() + "#" + town.getSpawnLocation().getY() + "#" + town.getSpawnLocation().getZ() : null);
            statement.setString(7, town.getMapColor());
            statement.executeUpdate();
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }


    private void sendResidentDataToDatabase(Resident resident) {
        if (resident == null) {
            System.err.println("Resident is null");
            return;
        }

        try (Connection connection = SQL.getCon();
             PreparedStatement statement = connection.prepareStatement(
                     "REPLACE INTO " + SchemeSQL.TABLE_PREFIX + "RESIDENTS (uuid, playername, town, nation, townranks, nationranks, language, permissionstown, permissionsnation) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)")) {
            statement.setString(1,resident.getUuid().toString());
            statement.setString(2, resident.getPlayerName());

            statement.setString(3, resident.getTown() != null ? resident.getTown().getName() : null);
            statement.setString(4, resident.getNation() != null ? resident.getNation().getName() : null);

            String rankTown = (resident.getRankTown() != null ? resident.getRankTown().getName() : null);
            statement.setString(5, rankTown);

            String rankNation = (resident.getRankNation() != null ? resident.getRankNation().getName() : null);
            statement.setString(6, rankNation);

            statement.setString(7, resident.getLanguage() != null && !resident.getLanguage().isEmpty() ? resident.getLanguage() : null);

            String permissions = (resident.getPermissionsTown() != null && !resident.getPermissionsTown().isEmpty()) ? "[" + resident.getPermissionsTown().stream().collect(Collectors.joining(", ")) + "]" : null;
            statement.setString(8, permissions);

            String permissions2 = (resident.getPermissionsNation() != null && !resident.getPermissionsNation().isEmpty()) ? "[" + resident.getPermissionsNation().stream().collect(Collectors.joining(", ")) + "]" : null;
            statement.setString(9, permissions2);

            statement.executeUpdate();
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }
}

File: src/main/kotlin/dev/loveeev/astraTowny/data/unload/UnLoad.kt
package dev.loveeev.astratowny.data.unload

import dev.loveeev.astratowny.AstraTowny
import dev.loveeev.astratowny.database.SQL
import dev.loveeev.astratowny.database.SchemeSQL
import dev.loveeev.astratowny.manager.TownManager
import dev.loveeev.astratowny.objects.Resident
import dev.loveeev.astratowny.objects.Town
import it.unimi.dsi.fastutil.objects.ObjectOpenHashSet
import org.bukkit.Location
import java.sql.SQLException
import java.util.logging.Level

class UnLoad {

    init {
        AstraTowny.instance.logger.info("Unloading...")
    }

    fun unLoadResident() {
        TownManager.residents.parallelStream().forEach { resident ->
                try {
                    SQL.getCon()?.use { connection ->
                        val query = "REPLACE INTO ${SchemeSQL.tablePrefix}RESIDENTS (uuid, playername, language, ranktown, ranknation) VALUES (?, ?, ?, ?, ?)"
                        connection.prepareStatement(query).use { statement ->
                            statement.setString(1, resident.uuid.toString())
                            statement.setString(2, resident.playerName)
                            statement.setString(3, resident.language)
                            statement.setString(4, resident.townRank?.name)
                            statement.setString(5, resident.nationRank?.name)
                            statement.executeUpdate()
                        }
                    }
                } catch (e: SQLException) {
                    AstraTowny.instance.logger.log(Level.SEVERE, "Error saving resident: ${resident.uuid}", e)
                }
            }
    }

    fun unloadPlots() {
        TownManager.towns.forEach { _, town ->
            town.plots.forEach { townBlock, plot ->
                SQL.getCon()?.use { connection ->
                    val query = "REPLACE INTO ${SchemeSQL.tablePrefix}PLOTS (uuid, townBlock, name, owner, residents, status, price) VALUES (?, ?, ?, ?, ?, ?, ?)"
                    connection.prepareStatement(query).use { statement ->
                        statement.setString(1, plot.uuid.toString())
                        statement.setString(2, townBlock.toString())
                        statement.setString(3, plot.name)
                        statement.setString(4, plot.owner?.playerName)
                        statement.setString(5, handleToStringListt(plot.residents))
                        statement.setString(6, plot.status.name)
                        statement.setDouble(7, plot.price)
                        statement.executeUpdate()
                    }
                }
            }
        }
    }
    /*
    fun unloadTownBlocks() {
        TownManager.townBlocks.values.parallelStream().forEach { townBlock ->
            try {
                SQL.getCon()?.use { connection ->
                    val query = "REPLACE INTO ${SchemeSQL.tablePrefix}TOWNBLOCKS (world, X, Z, town) VALUES (?, ?, ?, ?)"
                    connection.prepareStatement(query).use { statement ->
                        statement.setString(1, townBlock.world.name)
                        statement.setInt(2, townBlock.x)
                        statement.setInt(3, townBlock.z)
                        statement.setString(4, townBlock.town.name)
                        statement.executeUpdate()
                    }
                }
            } catch (e: SQLException) {
                AstraTowny.instance.logger.log(Level.SEVERE, "Error saving townBlock in ${townBlock.world} at (${townBlock.x}, ${townBlock.z})", e)
            }
        }
    }

     */



    fun unLoadTowns() {
        TownManager.towns.values.parallelStream().forEach { town ->
                try {
                    SQL.getCon()?.use { connection ->
                        val query = "REPLACE INTO ${SchemeSQL.tablePrefix}TOWNS (uuid, name, mayor, nation, homeblock, spawn, balance, mapcolor, residents) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)"
                        connection.prepareStatement(query).use { statement ->
                            statement.setString(1, town.uuid.toString())
                            statement.setString(2, town.name)
                            statement.setString(3, town.mayor?.playerName ?: "")
                            statement.setString(4, town.nation?.name ?: "")
                            statement.setString(5, town.homeBlock.toString())
                            statement.setString(6, handleToStringLocation(town.spawnLocation))
                            statement.setString(7, town.balance.toString())
                            statement.setString(8, town.mapColor)
                            statement.setString(9, handleToStringListt(town.residents))
                            statement.executeUpdate()
                        }
                    }
                } catch (e: SQLException) {
                    AstraTowny.instance.logger.log(Level.SEVERE, "Error saving town: ${town.uuid}", e)
                }
            }
    }

    private fun handleToStringListt(residents: ObjectOpenHashSet<Resident>): String {
        return residents.joinToString("#") { it.playerName }
    }


    private fun handleToStringLocation(spawnLocation: Location?): String? {
        return "${spawnLocation?.world?.name}#${spawnLocation?.x}#${spawnLocation?.y}#${spawnLocation?.z}#${spawnLocation?.yaw}#${spawnLocation?.pitch}"
    }

    fun unLoadNations() {
        TownManager.nations.values.parallelStream().forEach { nation ->
                try {
                    SQL.getCon()?.use { connection ->
                        val query = "REPLACE INTO ${SchemeSQL.tablePrefix}NATIONS (uuid, name, capital, balance, mapcolor, towns) VALUES (?, ?, ?, ?, ?, ?)"
                        connection.prepareStatement(query).use { statement ->
                            statement.setString(1, nation.uuid.toString())
                            statement.setString(2, nation.name)
                            statement.setString(3, nation.capital?.name ?: "")
                            statement.setString(4, nation.balance.toString())
                            statement.setString(5, nation.mapColor)
                            statement.setString(6, handleToStringList(nation.towns))
                            statement.executeUpdate()
                        }
                    }
                } catch (e: SQLException) {
                    AstraTowny.instance.logger.log(Level.SEVERE, "Error saving nation: ${nation.uuid}", e)
                }
            }
    }

    private fun handleToStringList(towns: ObjectOpenHashSet<Town>): String {
        return towns.joinToString("#") { it.name }
    }

}